File: graphs/cycles.cpp
#include<bits/stdc++.h>
using namespace std;

// 'u' is a node 'from', 'v' is a node 'to': from u to v
void add_edge_list(vector<vector<int>> &g, int u, int v) {
    g[u].push_back(v);      
    // g[v].push_back(u);      // Delete this line if graph is directed
}

void add_edge_list_undirected(vector<vector<int>> &g, int u, int v) {
    g[u].push_back(v);      
    g[v].push_back(u);      // Delete this line if graph is directed
}

// Check if the undirected graph has a cycle. DFS
bool has_cycle_visit(vector<vector<int>> &g, int u, bool* visited, int parent) {
    visited[u] = true;
    for (auto v: g[u]) {
        if (!visited[v]) 
            return has_cycle_visit(g, v, visited, u);
        else if (v != parent)
            return true;
    }
    return false;
}

bool has_cycle(vector<vector<int>> &g) {
    int n = g.size();
    bool* visited = new bool[n]();
    for (int i = 0; i < n; i++) {
        if (!visited[i]) 
            return has_cycle_visit(g, i, visited, -1);
    }
    return false;
}

// Check if the directed graph has a cycle. DFS
bool has_cycle_visit_directed(vector<vector<int>> g, int u, bool* visited, bool* recursion_stack) {
    if (!visited[u]) {
        visited[u] = true;
        recursion_stack[u] = true;

        for (auto v: g[u]) {
            if (!visited[v] && has_cycle_visit_directed(g, v, visited, recursion_stack))
                return true;
            else if (recursion_stack[v])
                return true;
        }

        recursion_stack[u] = false;
        return false;
    }
}

bool has_cycle_directed(vector<vector<int>> &g) {
    int n = g.size();

    bool* visited = new bool[n]();
    bool* recursion_stack = new bool[n]();

    for (int i = 0; i < n; i++) {
        return has_cycle_visit_directed(g, i, visited, recursion_stack);
    }

    return false;
}

int main() {
    int n, e; cin >> n >> e;

    // Adjacency list
    vector<vector<int>> adj_list(n, vector<int>());
    vector<vector<int>> adj_list2(n, vector<int>());

    // Our nodes are from 0 to n-1
    while (e--) {
        int u, v; cin >> u >> v;
        add_edge_list(adj_list, u, v);
        add_edge_list_undirected(adj_list2, u, v);
    }

    if (has_cycle_directed(adj_list))
        cout << "Has cycle.\n"; 
    else 
        cout << "Doesn't have cycle.\n";

    if (has_cycle(adj_list2)) 
        cout << "Has cycle.\n";
    else   
        cout << "Doesn't have cycle.\n";

    return 0;
}

File: graphs/cycles.py
def add_edge(graph, u, v):
    """
    Add an edge from node u to node v in the graph.
    """
    graph[u].append(v)


def add_edge_undirected(graph, u, v):
    """
    Add an undirected edge between node u and node v in the graph.
    """
    graph[u].append(v)
    graph[v].append(u)


def has_cycle_visit(graph, u, visited, parent):
    """
    Helper function to check if there is a cycle in an undirected graph using DFS.
    """
    visited[u] = True
    for v in graph[u]:
        if not visited[v]:
            if has_cycle_visit(graph, v, visited, u):
                return True
        elif v != parent:
            return True
    return False


def has_cycle(graph):
    """
    Check if the undirected graph has a cycle using DFS.
    """
    n = len(graph)
    visited = [False] * n
    for i in range(n):
        if not visited[i]:
            if has_cycle_visit(graph, i, visited, -1):
                return True
    return False


def has_cycle_visit_directed(graph, u, visited, recursion_stack):
    """
    Helper function to check if there is a cycle in a directed graph using DFS.
    """
    visited[u] = True
    recursion_stack[u] = True
    for v in graph[u]:
        if not visited[v]:
            if has_cycle_visit_directed(graph, v, visited, recursion_stack):
                return True
        elif recursion_stack[v]:
            return True
    recursion_stack[u] = False
    return False


def has_cycle_directed(graph):
    """
    Check if the directed graph has a cycle using DFS.
    """
    n = len(graph)
    visited = [False] * n
    recursion_stack = [False] * n
    for i in range(n):
        if not visited[i]:
            if has_cycle_visit_directed(graph, i, visited, recursion_stack):
                return True
    return False


if __name__ == "__main__":
    # Input number of nodes and edges
    n, e = map(int, input().split())

    # Adjacency list for the directed and undirected graphs
    adj_list = [[] for _ in range(n)]
    adj_list_undirected = [[] for _ in range(n)]

    # Adding edges to the graphs
    for _ in range(e):
        u, v = map(int, input().split())
        add_edge(adj_list, u, v)
        add_edge_undirected(adj_list_undirected, u, v)

    # Checking for cycles in directed and undirected graphs
    if has_cycle_directed(adj_list):
        print("Has cycle in directed graph.")
    else:
        print("Doesn't have cycle in directed graph.")

    if has_cycle(adj_list_undirected):
        print("Has cycle in undirected graph.")
    else:
        print("Doesn't have cycle in undirected graph.")
